siva/library.go

```diff
@@ -0,0 +1,244 @@
+	"github.com/src-d/go-borges/util"
+
+	billy "gopkg.in/src-d/go-billy.v4"
+	"gopkg.in/src-d/go-billy.v4/osfs"
+	butil "gopkg.in/src-d/go-billy.v4/util"
+	errors "gopkg.in/src-d/go-errors.v1"
+)
+
+// ErrLocationExists when the location to be created already exists.
+var ErrLocationExists = errors.NewKind("location %s already exists")
+
+// Library represents a borges.Library implementation based on siva files.
+type Library struct {
+	id            borges.LibraryID
+	fs            billy.Filesystem
+	tmp           billy.Filesystem
+	transactional bool
+	timeout       time.Duration
+	locReg        *locationRegistry
+}
+
+// LibraryOptions hold configuration options for the library.
+type LibraryOptions struct {
+	// Transactional enables transactions for repository writes.
+	Transactional bool
+	// Timeout is the time it will wait while another transaction
+	// is being done before error. 0 means default.
+	Timeout time.Duration
+	// RegistryCache is the maximum number of locations in the cache. A value
+	// of 0 disables the cache.
+	RegistryCache int
+	// TempFS is the temporary filesystem to do transactions and write files.
+	TempFS billy.Filesystem
+}
+
+var _ borges.Library = (*Library)(nil)
+
+// txTimeout is the default transaction timeout.
+const txTimeout = 60 * time.Second
+
+// NewLibrary creates a new siva.Library.
+func NewLibrary(
+	id string,
+	fs billy.Filesystem,
+	ops LibraryOptions,
+) (*Library, error) {
+	lr, err := newLocationRegistry(ops.RegistryCache)
+	if err != nil {
+		return nil, err
+	}
+
+	timeout := ops.Timeout
+	if timeout == 0 {
+		timeout = txTimeout
+	}
+
+	tmp := ops.TempFS
+	if tmp == nil {
+		dir, err := ioutil.TempDir("", "go-borges")
+		if err != nil {
+			return nil, err
+		}
+
+		tmp = osfs.New(dir)
+	}
+
+	return &Library{
+		id:            borges.LibraryID(id),
+		fs:            fs,
+		tmp:           tmp,
+		transactional: ops.Transactional,
+		timeout:       timeout,
+		locReg:        lr,
+	}, nil
+}
+
+// ID implements borges.Library interface.
+func (l *Library) ID() borges.LibraryID {
+	return l.id
+}
+
+// Init implements borges.Library interface.
+func (l *Library) Init(borges.RepositoryID) (borges.Repository, error) {
+	return nil, borges.ErrNotImplemented.New()
+}
+
+// Get implements borges.Library interface.
+func (l *Library) Get(repoID borges.RepositoryID, mode borges.Mode) (borges.Repository, error) {
+	ok, _, locID, err := l.Has(repoID)
+	if err != nil {
+		return nil, err
+	}
+
+	if !ok {
+		return nil, borges.ErrRepositoryNotExists.New(repoID)
+	}
+
+	loc, err := l.Location(locID)
+	if err != nil {
+		return nil, err
+	}
+
+	return loc.Get(repoID, mode)
+}
+
+// GetOrInit implements borges.Library interface.
+func (l *Library) GetOrInit(borges.RepositoryID) (borges.Repository, error) {
+	return nil, borges.ErrNotImplemented.New()
+}
+
+// TODO: find if we have to use ".git" suffix for repository ids
+func toRepoID(endpoint string) borges.RepositoryID {
```

Comment:
At the plain implementation is without: https://github.com/src-d/go-borges/blob/master/plain/location.go#L124